server/mailet/mailets/src/main/java/org/apache/james/transport/matchers/dlp/DlpDomainRules.java

```diff
@@ -0,0 +1,310 @@
+
+            private final Pattern pattern;
+
+            private RecipientsMatcher(Pattern pattern) {
+                this.pattern = pattern;
+            }
+
+            @Override
+            public boolean doTest(Mail mail) throws MessagingException, IOException {
+                return listRecipientsAsString(mail).anyMatch(pattern.asPredicate());
+            }
+
+            private Stream<String> listRecipientsAsString(Mail mail) throws MessagingException {
+                return Stream.concat(listEnvelopRecipients(mail), listHeaderRecipients(mail));
+            }
+
+            private Stream<String> listEnvelopRecipients(Mail mail) {
+                return mail.getRecipients().stream().map(MailAddress::asString);
+            }
+
+            private Stream<String> listHeaderRecipients(Mail mail) throws MessagingException {
+                return Optional.ofNullable(mail.getMessage())
+                    .flatMap(Throwing.function(m -> Optional.ofNullable(m.getAllRecipients())))
+                    .map(Rule::asStringStream)
+                    .orElse(Stream.of());
+            }
+
+        }
+
+        private static class SenderMatcher implements Rule.MatcherFunction {
+
+            private final Pattern pattern;
+
+            private SenderMatcher(Pattern pattern) {
+                this.pattern = pattern;
+            }
+
+            @Override
+            public boolean doTest(Mail mail) throws MessagingException {
+                return listSenders(mail).anyMatch(pattern.asPredicate());
+            }
+
+            private Stream<String> listSenders(Mail mail) throws MessagingException {
+                return Stream.concat(listEnvelopSender(mail), listFromHeaders(mail));
+            }
+
+            private Stream<String> listEnvelopSender(Mail mail) {
+                return OptionalUtils.toStream(Optional.ofNullable(mail.getSender()).map(MailAddress::asString));
+            }
+
+            private Stream<String> listFromHeaders(Mail mail) throws MessagingException {
+                MimeMessage message = mail.getMessage();
+                if (message != null) {
+                    return asStringStream(message.getFrom());
+                }
+                return Stream.of();
+            }
+
+        }
+
+        private final DLPConfigurationItem.Id id;
+        private final MatcherFunction matcher;
+
+        public Rule(DLPConfigurationItem.Id id, MatcherFunction matcher) {
+            this.id = id;
+            this.matcher = matcher;
+        }
+
+        public DLPConfigurationItem.Id id() {
+            return id;
+        }
+
+        public boolean match(Mail mail) {
+            return matcher.test(mail);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (o instanceof Rule) {
+                Rule other = (Rule) o;
+                return Objects.equals(id, other.id) &&
+                    Objects.equals(matcher, other.matcher);
```

Comment:
matcher `MatcherFunction` implementations (`SenderMatcher`, `RecipientMatcher`, `ContentMatcher`) are missing override `equals()` & `hashCode()`

Reply:
the problem is that Pattern doesn't implement equals/hashCode anyway.
BTW, not implementing equals/hashcode is a valid implementation as it does instance equality anyway.
Do you think we should work around Pattern limitation ? What would we gain then ?

Reply:
I see a `private final ImmutableCollection<Rule> rules` in side `DlpDomainRules`. then `Rule` should override `equals()` & `hashCode()`.  But if we don't specify Matcher `equals()` & `hashCode()` then `equals()` & `hashCode()` of `Rule` are useless. 
`Pattern.toString()` with me is okay to use in `MatcherFunction` implementation. E.g:

```java
            @Override
            public final boolean equals(Object o) {
                if (o instanceof SenderMatcher) {
                    SenderMatcher that = (SenderMatcher) o;
                    return Objects.equals(this.pattern.toString(), that.pattern.toString());
                }
                return false;
            }

            @Override
            public final int hashCode() {
                return Objects.hash(pattern.toString());
            }
```

Reply:
About objects in a Collection : a collection is just a set of items. It doesn't rely on equals/hashcode in any way. List doesn't, neither.
You have to start worrying about that in Set and when you use it as Map key. It's not the case here.
Even then, default hashcode/equals are ok. They just don't ensure "content equality" but only "reference equality".
I'm not against using `Pattern.pattern()` in equals implementation but still, I'm not sure what we would gain.